### PR TITLE
meta(sync): tracing

### PR DESCRIPTION
**Pull request by rkrasiuk**
This PR adds tracing to sync stages. The target for the traces/logs is written in a format `sync::stages::<stage_name>`. The pipeline progress logic is written at the `info` level and per data block logs are written at the `trace` level

### CODE REVIEW

File: crates/stages/src/stages/bodies.rs [hunk 1]
```diff
@@ -171,6 +172,7 @@ impl<DB: Database, D: BodyDownloader, C: Consensus> Stage<DB> for BodyStage<D, C
         // - We reached our target and the target was not limited by the batch size of the stage
         let capped = target < previous_stage_progress;
         let done = highest_block < target || !capped;
+        info!(target : "sync::stages::bodies", stage_progress = highest_block, done = done, "Finished stage.");
```
Comment (onbjerg) [inline crates/stages/src/stages/bodies.rs:175]: not entirely, should maybe say "synced until" or smth since it will sync over multiple runs

Comment (rkrasiuk) [reply crates/stages/src/stages/bodies.rs]: changed to `Sync iteration finished`

File: crates/stages/Cargo.toml [hunk 2]
```diff
@@ -41,3 +41,6 @@ tokio-stream = "0.1.10"
 tempfile = "3.3.0"
 assert_matches = "1.5.0"
 rand = "0.8.5"
+# TODO:
+env_logger = "*"
+test-log = "*"
```
Comment (onbjerg) [inline crates/stages/Cargo.toml:46]: What's this for?

File: crates/stages/src/stages/bodies.rs [hunk 3]
```diff
@@ -175,19 +175,22 @@ impl<DB: Database, D: BodyDownloader, C: Consensus> Stage<DB> for BodyStage<D, C
             // Increment the transition if the block contains an addition block reward.
             // If the block does not have a reward, the transition will be the same as the
             // transition at the last transaction of this block.
-            if self.consensus.has_block_reward(block_number) {
+            let has_reward = self.consensus.has_block_reward(numhash.number());
+            debug!(target: "sync::stages::bodies", has_reward, ?numhash, "Block reward");
```
Comment (onbjerg) [inline crates/stages/src/stages/bodies.rs:179]:
```suggestion
            trace!(target: "sync::stages::bodies", has_reward, ?numhash, "Block reward");
```

File: crates/stages/src/stages/bodies.rs [hunk 4]
```diff
@@ -107,20 +108,17 @@ impl<DB: Database, D: BodyDownloader, C: Consensus> Stage<DB> for BodyStage<D, C
...
 
-        // Get id for the first transaction in the block
+        // Get id for the first transaction and first transition in the block
         let (mut current_tx_id, mut transition_id) = tx.get_next_block_ids(starting_block)?;
 
         // NOTE(onbjerg): The stream needs to live here otherwise it will just create a new iterator
         // on every iteration of the while loop -_-
         let mut bodies_stream = self.downloader.bodies_stream(bodies_to_download.iter());
         let mut highest_block = stage_progress;
+        trace!(target: "sync::stages::bodies", highest_block, tx_id = current_tx_id, transition_id, "Commencing sync");
```
Comment (onbjerg) [inline crates/stages/src/stages/bodies.rs:118]:
```suggestion
        trace!(target: "sync::stages::bodies", stage_progress, target, start_tx_id = current_tx_id, transition_id, "Commencing sync");
```

File: crates/stages/src/stages/bodies.rs [hunk 5]
```diff
@@ -175,19 +175,22 @@ impl<DB: Database, D: BodyDownloader, C: Consensus> Stage<DB> for BodyStage<D, C
...
-            highest_block = block_number;
+            highest_block = numhash.number();
         }
 
         // The stage is "done" if:
         // - We got fewer blocks than our target
         // - We reached our target and the target was not limited by the batch size of the stage
         let capped = target < previous_stage_progress;
         let done = highest_block < target || !capped;
+        info!(target: "sync::stages::bodies", stage_progress = highest_block, done, "Sync iteration finished");
```
Comment (onbjerg) [inline crates/stages/src/stages/bodies.rs:193]:
```suggestion
        info!(target: "sync::stages::bodies", stage_progress = highest_block, target, done, "Sync iteration finished");
```

File: crates/stages/src/stages/headers.rs [hunk 6]
```diff
@@ -107,32 +98,23 @@ impl<DB: Database, D: HeaderDownloader, C: Consensus, H: HeadersClient, S: Statu
...
-                        );
+                        warn!(target: "sync::stages::headers", "No response for header request");
                         return Err(StageError::Recoverable(DownloadError::Timeout.into()))
                     }
                     DownloadError::HeaderValidation { hash, error } => {
-                        error!(
-                            target: "sync::stages::headers",
-                            "Validation error for header {hash}: {error}"
-                        );
+                        error!( target: "sync::stages::headers", ?error, ?hash, "Validation error");
```
Comment (onbjerg) [inline crates/stages/src/stages/headers.rs:105]:
```suggestion
                        error!(target: "sync::stages::headers", ?error, ?hash, "Validation error");
```

File: crates/stages/src/stages/headers.rs [hunk 7]
```diff
@@ -107,32 +98,23 @@ impl<DB: Database, D: HeaderDownloader, C: Consensus, H: HeadersClient, S: Statu
...
-                        );
+                        error!(target: "sync::stages::headers", ?error, "Unexpected error");
                         return Err(StageError::Recoverable(error.into()))
                     }
                 },
             }
         }
 
         // Write total difficulty values after all headers have been inserted
+        info!(target: "sync::stages::headers", head = ?head.hash(), "Writing total difficulty");
```
Comment (onbjerg) [inline crates/stages/src/stages/headers.rs:117]:
```suggestion
        debug!(target: "sync::stages::headers", head = ?head.hash(), "Writing total difficulty");
```

also we should have a "finished iteration" like in the bodies stage

File: crates/stages/src/stages/senders.rs [hunk 8]
```diff
@@ -63,6 +64,7 @@ impl<DB: Database> Stage<DB> for SendersStage {
         let max_block_num = previous_stage_progress.min(stage_progress + self.commit_threshold);
 
         if max_block_num <= stage_progress {
+            info!(target: "sync::stages::senders", max_block_num, stage_progress, "Target block already reached");
```
Comment (onbjerg) [inline crates/stages/src/stages/senders.rs:67]:
```suggestion
            info!(target: "sync::stages::senders", target =max_block_num, stage_progress, "Target block already reached");
```

File: crates/stages/src/stages/senders.rs [hunk 9]
```diff
@@ -88,24 +91,27 @@ impl<DB: Database> Stage<DB> for SendersStage {
             .take_while(|res| res.as_ref().map(|(k, _)| *k <= end_tx_index).unwrap_or_default());
 
         // Iterate over transactions in chunks
+        info!(target: "sync::stages::senders", start_tx_index, end_tx_index, batch_size = self.batch_size, "Recovering senders");
```
Comment (onbjerg) [inline crates/stages/src/stages/senders.rs:94]:
```suggestion
        info!(target: "sync::stages::senders", start_tx_index, end_tx_index, "Recovering senders");
```

**Review by onbjerg — APPROVED** [review]
Minor nits, we should try to keep the labels consistent (e.g. `stage_progress`, `target` and so on)

File: crates/stages/Cargo.toml [hunk 10]
```diff
@@ -41,3 +41,6 @@ tokio-stream = "0.1.10"
 tempfile = "3.3.0"
 assert_matches = "1.5.0"
 rand = "0.8.5"
+# TODO:
+env_logger = "*"
```
Comment (mattsse) [inline crates/stages/Cargo.toml:45]: should either use reth-tracing crate and/or subscriber for test logging

File: crates/stages/src/stages/bodies.rs [hunk 11]
```diff
@@ -352,7 +355,7 @@ mod tests {
     }
 
     /// Same as [full_body_download] except we have made progress before
-    #[tokio::test]
+    #[test_log::test(tokio::test)]
```
Comment (mattsse) [inline crates/stages/src/stages/bodies.rs:358]: I use ` reth_tracing::init_tracing();` on tests where tracing is useful.

**Review by mattsse — CHANGES_REQUESTED** [review]

File: crates/stages/src/stages/bodies.rs (hunk 11 above)
Comment (rkrasiuk) [reply crates/stages/src/stages/bodies.rs:358]: sg, just wanted to test some things, forgot to remove

**Review by mattsse — APPROVED** [review]
lgtm,

good to merge?

**Pull request merged**